kolena/_experimental/classification/utils.py

```diff
@@ -125,3 +154,98 @@ def compute_confusion_matrix(
+    threshold_idxs = np.r_[distinct_value_indices, y_true.size - 1]
+    # Compute the cumulative sums of true positives and false positives
+    tps = np.cumsum(y_true)[threshold_idxs]
+    fps = np.cumsum(1 - y_true)[threshold_idxs]
+    # Drop collinear points (copied from sklearn)
+    if len(fps) > 2:
+        optimal_indices = np.where(np.r_[True, np.logical_or(np.diff(fps, 2), np.diff(tps, 2)), True])[0]
+        fps = fps[optimal_indices]
+        tps = tps[optimal_indices]
+    fps = np.r_[0, fps]
+    tps = np.r_[0, tps]
+    # Map cumulative sums to tpr and fpr
+    if fps[-1] <= 0:
+        # No negative samples in y_true, false positive value should be meaningless
+        fpr = []
+    else:
+        fpr = fps / fps[-1]
+        fpr = fpr.tolist()
+    if tps[-1] <= 0:
+        # No positive samples in y_true, true positive value should be meaningless
+        tpr = []
+    else:
+        tpr = tps / tps[-1]
+        tpr = tpr.tolist()
+    return fpr, tpr
+
+
+def compute_roc_curves(
+    ground_truths: List[Optional[ClassificationLabel]],
+    inferences: List[List[ScoredClassificationLabel]],
+    labels: Optional[List[str]] = None,
+    title: Optional[str] = None,
+) -> Optional[CurvePlot]:
+    """
+    Computes OvR (one-vs-rest) ROC (receiver operating characteristic) curves for each class appears in `ground_truths`
+    if not specified.
+
+    :param ground_truths: The ground truth classification labels. For binary classification, the negative class is
+        `None`.
+    :param inferences: The inferences. For `N`-class problems, `label` is expected to contain `N` entries, one for
+        each class and its associated confidence score.
+    :param labels: The labels to plot. If not specified, classes appear in `ground_truths` are used. Use `labels` to
+        specify the evaluating classes especially if `ground_truths` only have negative classes.
+    :param title: The title of the plot.
+    :return: A `CurvePlot` if there is any valid `Curve` computed; otherwise, `None`.
+    """
+    if len(ground_truths) != len(inferences):
+        log.warn(
+            f"ground_truths ({len(ground_truths)}) and inferences ({len(inferences)}) differ in length",
+        )
+        return None
+
+    if len(ground_truths) <= 2:
+        log.warn(
+            f"insufficient # of samples to compute a roc curve — need at least 3 but received {len(ground_truths)}",
+        )
+        return None
+
+    if labels is None:
+        labels = sorted({gt.label for gt in ground_truths if gt is not None})
+
+    curves: List[Curve] = []
+    for label in labels:
+        y_true = [1 if gt is not None and gt.label == label else 0 for gt in ground_truths]
+        y_score = [get_label_confidence(label, inf) for inf in inferences]
+        fpr_values, tpr_values = _roc_curve(y_true=y_true, y_score=y_score)
+
+        if len(fpr_values) > 0 and len(tpr_values) > 0 and len(fpr_values) == len(tpr_values):
```

Comment:
To my understanding, this identifies valid plot data. Is it useful to give users feedback for invalid path, or is `invalid` path a normal condition?

Reply:
Yeah here the invalid path is when there is no negative sample or no positive sample in a test case.  IMO it's pretty easy to spot such cases by looking at the per class metric (`n_images`) which is why I decided not to log anything.